v2/internal/genericarmclient/policy_register_rp.go

```diff
@@ -0,0 +1,221 @@
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	armpolicy "github.com/Azure/azure-sdk-for-go/sdk/azcore/arm/policy"
+	armruntime "github.com/Azure/azure-sdk-for-go/sdk/azcore/arm/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	azpolicy "github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+
+	"github.com/Azure/azure-service-operator/v2/internal/version"
+)
+
+// NewRPRegistrationPolicy creates a policy object configured using the specified options.
+// The policy controls whether an unregistered resource provider should automatically be
+// registered. See https://aka.ms/rps-not-found for more information.
+func NewRPRegistrationPolicy(cred azcore.TokenCredential, o *azpolicy.ClientOptions) (azpolicy.Policy, error) {
+	conf, err := getConfiguration(o)
+	if err != nil {
+		return nil, err
+	}
+	authPolicy := armruntime.NewBearerTokenPolicy(cred, &armpolicy.BearerTokenOptions{Scopes: []string{conf.Audience + "/.default"}})
+	p := &rpRegistrationPolicy{
+		endpoint: conf.Endpoint,
+		pipeline: runtime.NewPipeline("generic", version.BuildVersion, runtime.PipelineOptions{PerRetry: []azpolicy.Policy{authPolicy}}, o),
+		options:  o,
+	}
+	return p, nil
+}
+
+type rpRegistrationPolicy struct {
+	endpoint string
+	pipeline runtime.Pipeline
+	options  *azpolicy.ClientOptions
+}
+
+func (r *rpRegistrationPolicy) Do(req *azpolicy.Request) (*http.Response, error) {
+	const unregisteredRPCode = "MissingSubscriptionRegistration"
+	var rp string
+	var resp *http.Response
+	var err error
+	// make the original request
+	resp, err = req.Next()
+	// getting a 409 is the first indication that the RP might need to be registered, check error response
+	if err != nil || resp.StatusCode != http.StatusConflict {
+		return resp, err
+	}
+	var reqErr requestError
+	if err = runtime.UnmarshalAsJSON(resp, &reqErr); err != nil {
+		return resp, err
+	}
+	if reqErr.ServiceError == nil {
+		return resp, errors.New("missing error information")
+	}
+	if !strings.EqualFold(reqErr.ServiceError.Code, unregisteredRPCode) {
+		// not a 409 due to unregistered RP
+		return resp, err
+	}
+	// RP needs to be registered.  start by getting the subscription ID from the original request
+	subID, err := getSubscription(req.Raw().URL.Path)
+	if err != nil {
+		return resp, err
+	}
+	// now get the RP from the error
+	rp, err = getProvider(reqErr)
+	if err != nil {
+		return resp, err
+	}
+	// create client and make the registration request
+	// we use the scheme and host from the original request
+	rpOps := &providersOperations{
+		p:     r.pipeline,
+		u:     r.endpoint,
+		subID: subID,
+	}
+	if _, err = rpOps.Register(req.Raw().Context(), rp); err != nil {
+		return resp, err
+	}
+	return resp, fmt.Errorf("registering Resource Provider %s with subscription. Try again later", rp)
+}
+
+func getSubscription(path string) (string, error) {
+	parts := strings.Split(path, "/")
+	for i, v := range parts {
```

Comment:
Can use regex matcher here?

Reply:
Could, but since I copied this code from the Go SDK I think it makes sense to (as much as possible) keep it similar so that we can easily apply changes if we want.

I copied their code rather than using their code because theirs forces retries and a wait on you (or else you get an ugly timeout error), and I wanted our users to see a better error than that.

Reply:
I'm not sure it would be easy to match with a Go regex anyway - isn't the limitation on prior-matches one of the issues?